### PR TITLE
Criteria pagination

### LINKED ISSUE #2707: Pageable and Criteria API - incorrect totalSize for many-to-many relations
### Expected Behavior

`totalSize` should be correct, counting only the final records.

### Actual Behaviour

When response has only 1 member, `totalSize` still shows value of 2, when that one member has 2 values from join.

Probably the easiest fix would be to count only distinct ID-s.

### Steps To Reproduce

Entities:
```kt
@MappedEntity
data class Author(
    @field:Id
    val id: Long,
    val name: String,
    @Relation(value = Relation.Kind.MANY_TO_MANY)
    val genres: List<Genre>
)

@MappedEntity
data class Genre(
    @field:Id
    val id: Long,
    val name: String
)
```

Create interface:
```kt
@JdbcRepository(dialect = Dialect.POSTGRES)
@Join(value = "genres", type = Join.Type.LEFT_FETCH)
interface AuthorRepository : CrudRepository<Author, Long>, JpaSpecificationExecutor<Author>
```

Create tables with test data:
```sql
create table author
(
    id   bigserial primary key,
    name text not null
);

insert into author (name) values ('Stephen King');

create table genre
(
    id   bigserial primary key,
    name text not null
);

insert into genre (name) values ('Horror'), ('Thriller'), ('Comedy');

create table author_genre
(
    author_id bigint not null references author (id) on delete cascade,
    genre_id  bigint not null references genre (id) on delete cascade,
    unique (author_id, genre_id)
);

insert into author_genre (author_id, genre_id) values (1, 1), (1, 2);
```

Do test: 
```kt
@MicronautTest
class TotalSizeIssueTest(
    private val authorRepository: AuthorRepository
) : StringSpec({

    "test totalSize should match the number of authors" {
        val response = authorRepository.findAll(where {
            val genresJoin =
                root.joinList<Author, Genre>("genres", JoinType.LEFT)

            or {
                genresJoin[Genre::name] eq "Horror"
                genresJoin[Genre::name] eq "Thriller"
            }
        }, Pageable.from(0, 10))

        response.totalSize shouldBe response.content.size
    }
})
```

The sizes should match, but the assertion fails, as response size is 1 but total size is 2.

### Environment Information

- Kotlin 1.9.21
- Micronaut 3.10.1

### Example Application

_No response_

### Version

3.10.1

### PR DESCRIPTION
**Pull request by dstepanov**
Fixes https://github.com/micronaut-projects/micronaut-data/issues/2707

Splitting criteria execution into two parts:
- First loading IDs of the entity
- Second fetching full entity content

To do so I had to implement support for Jakarta Tuple, because we need to fetch IDs and every ordered property.
Criteria interceptors are refactored to always use `CriteriaRepositoryOperations`, simplifying invocation for Micronaut/Hibernate criteria. Criteria now always include any kind of pagination / sorting into the query.

### CODE REVIEW

**Comment by dstepanov** [comment]
I will add docs later. It looks to me that the Criteria section of docs can be moved to the common section.

File: data-runtime/src/main/java/io/micronaut/data/runtime/operations/internal/sql/DefaultSqlPreparedQuery.java [hunk 1]
```diff
@@ -327,6 +356,9 @@ public List<Cursor> createCursors(List<Object> results, Pageable pageable, Runti
         if (pageable.getMode() == Mode.CURSOR_PREVIOUS) {
             Collections.reverse(results);
         }
+        RuntimePersistentEntity<E> persistentEntity = getPersistentEntity();
```
Comment (radovanradic) [inline data-runtime/src/main/java/io/micronaut/data/runtime/operations/internal/sql/DefaultSqlPreparedQuery.java]: `runtimePersistentEntity` is already passed and used below, not sure if this is needed and it looks confusing

**Comment by radovanradic** [comment]
Don't know if Sonar issues need to be addressed, some potential NPEs etc.

**Review by radovanradic — APPROVED** [review]

File: data-runtime/src/main/java/io/micronaut/data/runtime/operations/internal/sql/DefaultSqlPreparedQuery.java (hunk 1 above)
Comment (dstepanov) [reply data-runtime/src/main/java/io/micronaut/data/runtime/operations/internal/sql/DefaultSqlPreparedQuery.java]: Fixed

**Comment by dstepanov** [comment]
Addressed some of the Sonar issues

**Pull request merged**